src/transformers/generation/candidate_generator.py

```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        convert_kwargs = {"src": self.target_tokenizer, "dest": self.assistant_tokenizer}
+        remove_from_pkv = 0
+
+        if self.prev_tokens is None:
+            assistant_input_ids = self.convert_token_ids(input_ids, **convert_kwargs)
+            self.prev_target_ids = input_ids
+            self.prev_assistant_ids = assistant_input_ids
+            new_cur_len = assistant_input_ids.shape[-1]
+        else:
+            # input_ids contains all target prompt input ids and some new target input ids
+            if self.prev_target_ids.shape[1] > self.target_lookbehind:
+                start_index_in_target_window = self.prev_target_ids.shape[1] - self.target_lookbehind
+
+                new_assistant_ids = self.convert_token_ids(
+                    input_ids[:, start_index_in_target_window:], **convert_kwargs
+                )
+                prompt_use_length = new_assistant_ids.shape[1]
+                prompt_use = self.prev_assistant_ids[:, -prompt_use_length:]
+
+                replace_tokens_from_prompt, disrep_length, new_tokens_with_disrep, new_tokens_only, discrep_only = (
+                    _get_tokens_diag(prompt_use, new_assistant_ids)
+                )
+                assistant_input_ids = self.prev_assistant_ids
+
+                if new_tokens_only is not None:
+                    if disrep_length > 0 and discrep_only.shape[1] > 0:
+                        if disrep_length == discrep_only.shape[1]:
+                            assistant_input_ids[:, -disrep_length:] = discrep_only
+
+                        elif disrep_length > discrep_only.shape[1]:
+                            disrep_length_diff = disrep_length - discrep_only.shape[1]
+                            assistant_input_ids = assistant_input_ids[:, :-disrep_length_diff]
+                            assistant_input_ids[:, -discrep_only.shape[1] :] = discrep_only
+
+                        remove_from_pkv = disrep_length
+
+                    if new_tokens_only.shape[1] > 0:
+                        assistant_input_ids = torch.cat([assistant_input_ids, new_tokens_only], dim=-1)
+                else:
+                    # edge case: in case of no intersection between prompt and new_assistant_ids
+                    assistant_input_ids = torch.cat([assistant_input_ids, new_assistant_ids], dim=-1)
+
+                self.prev_assistant_ids = assistant_input_ids
+            else:
+                assistant_input_ids = self.convert_token_ids(input_ids, **convert_kwargs)
+                assistant_input_ids = torch.cat(
+                    [self.prev_assistant_ids, assistant_input_ids[:, self.prev_assistant_ids.shape[1] :]], dim=1
+                )
+
+            new_cur_len = assistant_input_ids.shape[-1]
+
+        max_new_tokens = int(self.num_assistant_tokens)
+
+        min_new_tokens = max(min(max_new_tokens, self.main_model_min_length - new_cur_len), 0)
+        if max_new_tokens == 0:
+            return input_ids, None
```

Comment:
We can move this block (except `min_new_tokens = ...`) to the top of `get_candidates`, so we can hit this return more quickly 😎